atom/browser/net/atom_network_delegate.cc

```diff
@@ -0,0 +1,388 @@
+      if (pattern.MatchesURL(url))
+        return true;
+    return false;
+  }
+
+  return true;
+}
+
+scoped_ptr<base::DictionaryValue> ExtractRequestInfo(net::URLRequest* request) {
+  scoped_ptr<base::DictionaryValue> dict(new base::DictionaryValue());
+  dict->SetInteger("id", request->identifier());
+  dict->SetString("url", request->url().spec());
+  dict->SetString("method", request->method());
+  content::ResourceType resourceType = content::RESOURCE_TYPE_LAST_TYPE;
+  auto info = content::ResourceRequestInfo::ForRequest(request);
+  if (info)
+    resourceType = info->GetResourceType();
+  dict->SetString("resourceType", ResourceTypeToString(resourceType));
+  dict->SetDouble("timestamp", base::Time::Now().ToDoubleT() * 1000);
+
+  return dict.Pass();
+}
+
+base::DictionaryValue* GetRequestHeadersDict(
+    const net::HttpRequestHeaders& headers) {
+  base::DictionaryValue* header_dict = new base::DictionaryValue();
+  net::HttpRequestHeaders::Iterator it(headers);
+  while (it.GetNext())
+    header_dict->SetString(it.name(), it.value());
+  return header_dict;
+}
+
+base::DictionaryValue* GetResponseHeadersDict(
+    const net::HttpResponseHeaders* headers) {
+  base::DictionaryValue* header_dict = new base::DictionaryValue();
+  if (headers) {
+    void* iter = nullptr;
+    std::string key;
+    std::string value;
+    while (headers->EnumerateHeaderLines(&iter, &key, &value))
+      header_dict->SetString(key, value);
+  }
+  return header_dict;
+}
+
+void OnBeforeURLRequestResponse(
+    const net::CompletionCallback& callback,
+    GURL* new_url,
+    const AtomNetworkDelegate::BlockingResponse& result) {
+  if (!result.redirectURL.is_empty())
+    *new_url = result.redirectURL;
+  callback.Run(result.Cancel());
+}
+
+void OnBeforeSendHeadersResponse(
+    const net::CompletionCallback& callback,
+    net::HttpRequestHeaders* headers,
+    const AtomNetworkDelegate::BlockingResponse& result) {
+  if (!result.requestHeaders.IsEmpty())
+    *headers = result.requestHeaders;
+  callback.Run(result.Cancel());
+}
+
+void OnHeadersReceivedResponse(
+    const net::CompletionCallback& callback,
+    const net::HttpResponseHeaders* original_response_headers,
+    scoped_refptr<net::HttpResponseHeaders>* override_response_headers,
+    const AtomNetworkDelegate::BlockingResponse& result) {
+  if (result.responseHeaders.get()) {
+    *override_response_headers = new net::HttpResponseHeaders(
+        original_response_headers->raw_headers());
+    void* iter = nullptr;
+    std::string key;
+    std::string value;
+    while (result.responseHeaders->EnumerateHeaderLines(&iter, &key, &value)) {
+      (*override_response_headers)->RemoveHeader(key);
+      (*override_response_headers)->AddHeader(key + ": " + value);
+    }
+  }
+  callback.Run(result.Cancel());
+}
+
+}  // namespace
+
+// static
+std::map<AtomNetworkDelegate::EventTypes, AtomNetworkDelegate::ListenerInfo>
+AtomNetworkDelegate::event_listener_map_;
```

Comment:
It should be a member data, a multi-session app can have multiple `NetworkDelegate`s.